### PR TITLE
Update dependencies for 5.7.0

### PR DESCRIPTION
**Pull request by str4d**
This PR also adds a CI check that enforces an allow-list for Rust dependency licenses.

### CODE REVIEW

**Comment by nuttycom** [comment]
[force-pushed](https://github.com/zcash/zcash/compare/87b45299758dd1f6e21581347ccd3d9a603f54f3..0cdb76e1e5d9049a260f26958db1ccbc7cab052f) to incorporate the mempool_packages nondeterministic CI failure fix.

File: Cargo.lock
Comment (nuttycom) [inline Cargo.lock:2444]: This looks like it's introducing a duplicate crate dependency.

**Review by nuttycom — CHANGES_REQUESTED** [review]
It looks like this introduces a duplicate dependency with differing versions for the `windows-target` crate.

File: Cargo.lock
Comment (str4d) [reply Cargo.lock:2444]: The only crate still depending on this older version is `dirs-sys 0.4.0`. The transitive dependency was updated in `dirs-sys 0.4.1` which would remove the duplicate dependency; however, that version also introduces a dependency on `option-ext` which is under the Mozilla Public License, a (weak) copyleft license that we currently do not have any dependency on (per the license checking logic I added in this PR). We would need to consider whether that is permissible for `zcashd` before upgrading.

Comment (str4d) [reply Cargo.lock:2444]: Also note that `terminfo` depends transitively on `dirs-sys 0.3`, which in turn depends on the `winapi` crate (which was the community-maintained API bindings to the Windows SDK before Microsoft started releasing their own API bindings under the `windows-sys` crate). So we already have effectively duplicate dependencies here, and they aren't going away any time soon (though hopefully over time).

Comment (daira) [reply Cargo.lock:2444]: Filed https://github.com/dirs-dev/dirs-sys-rs/issues/23, but I'm not hopeful since the similar https://github.com/dirs-dev/dirs-sys-rs/issues/21 was closed as wontfix. The maintainer apparently does not care that they made an incompatible license change in a supposedly non-breaking release.

Comment (daira) [reply Cargo.lock:2444]: Also filed https://github.com/dirs-dev/dirs-sys-rs/pull/24 to fix it.

**Comment by daira** [comment]
As discussed in standup today, I would prefer that we explicitly whitelist particular (crate, license) combinations for non-MIT/Apache Rust dependencies. That is no additional effort because I am already checking those dependencies in order to keep the debian/contrib/copyright` file up-to-date. In fact it's less effort, because it automates the task of finding out when the `copyright` file should be updated, even if it doesn't automate the update itself (which it probably shouldn't).

File: deny.toml
Comment (daira) [inline deny.toml:36]: I think this isn't quite accurate. See https://github.com/zcash/zcash/blob/4ca229cb8949807a46737109ea54a4afa6c33195/contrib/debian/copyright#L1503-L1705

File: contrib/debian/copyright
Comment (daira) [inline contrib/debian/copyright:243]: This is a good change.

File: deny.toml
Comment (daira) [reply deny.toml:36]: BTW I know that this is literally [used as an example for `[[licenses.clarify]]`](https://embarkstudios.github.io/cargo-deny/checks/licenses/cfg.html#example), but I think the example is wrong.

**Comment by daira** [comment]
I verified that https://github.com/zcash/zcash/pull/6731/commits/0cdb76e1e5d9049a260f26958db1ccbc7cab052f fixes #6698 and that https://github.com/zcash/zcash/pull/6731/commits/382ac84304118de4c207d126c86c5de25eb52c89 fixes #6702, as claimed.

File: deny.toml
Comment (str4d) [reply deny.toml:36]: That looks correct to me. There is some code under each of the licenses, so we need to accept all three:
- MIT
  - `The code in third_party/fiat carries the MIT license`
- ISC
  - `*ring* uses an ISC-style license like BoringSSL for code in new files, including in particular all the Rust code`
  - `BoringSSL [..] [f]iles that are completely new have a Google copyright and an ISC license.`
- OpenSSL
  - `BoringSSL is a fork of OpenSSL. As such, large parts of it fall under OpenSSL 
  licensing`

Incidentally, this `[[licenses.clarify]]` section for `ring` is taken directly from the [`cargo-deny` example config](https://github.com/EmbarkStudios/cargo-deny/blob/9ca5722c3f34ed8cd6a59d6bff0c32325be9a2c5/deny.template.toml#L142-L160).

**Comment by str4d** [comment]
[Rebased on current `master`](https://github.com/zcash/zcash/compare/0cdb76e1e5d9049a260f26958db1ccbc7cab052f..19fdace9dcd897a431d775a53cc0ceacebffba07).

**Comment by str4d** [comment]
[Force-pushed](https://github.com/zcash/zcash/compare/19fdace9dcd897a431d775a53cc0ceacebffba07..a803af218642e4481de4e3acdfe77239b8ed22ec) to require explicit exceptions for Rust dependencies not licensed as either `MIT` or `Apache-2.0`.

File: deny.toml
Comment (str4d) [reply deny.toml:47]: > BTW I know that this is literally [used as an example for `[[licenses.clarify]]`](https://embarkstudios.github.io/cargo-deny/checks/licenses/cfg.html#example), but I think the example is wrong.

The `ring` exception example can be traced back to their README when it was [added in 2019](https://github.com/EmbarkStudios/cargo-deny/pull/34/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R103-R114) (as part of updating the README after exceptions were added). Prior to that it was used as an example of skipping a check (since [the initial README](https://github.com/EmbarkStudios/cargo-deny/commit/78b676c01dd7c62f04d1c28806ee3f9a8eb18adb)).

Comment (str4d) [reply deny.toml:47]: On the flipside, I just noticed https://github.com/EmbarkStudios/cargo-deny/issues/310 and then went looking in the `ring` repo, and they use `LicenseRef-ring`: https://github.com/briansmith/ring/blob/95948b3977013aed16db92ae32e6b8384496a740/deny.toml#L18-L23

So I'll defer to them and match.

**Comment by str4d** [comment]
[Force-pushed](https://github.com/zcash/zcash/compare/a803af218642e4481de4e3acdfe77239b8ed22ec..f37cd195563da0c45d5d5c6d77c4d086e15dc949) to use the same license reference as the upstream `ring` crate.

I also realised while doing so that the `deny.toml` is not pinning the `orchard` crate's `COPYING` file (which is where the Zcash exception is specified). However it turns out this is not currently possible to pin, due to a bug in `cargo-deny`. I've opened an issue for it: https://github.com/EmbarkStudios/cargo-deny/issues/524

File: Cargo.lock
Comment (str4d) [reply Cargo.lock:2444]: Issue and PR were silently closed. That's the developer's choice and they are welcome to it, but I'll start the process of migrating us away from this dependency.

Comment (str4d) [reply Cargo.lock:2444]: Opened https://github.com/zcash/librustzcash/pull/864.

**Comment by str4d** [comment]
[Force-pushed](https://github.com/zcash/zcash/compare/f37cd195563da0c45d5d5c6d77c4d086e15dc949..e8ad08aed185eeb06b18eff715f1d69237107e63) to bump to `cargo-vet 0.8` and update dependencies again to include https://github.com/zcash/librustzcash/pull/864.

File: Cargo.lock
Comment (str4d) [reply Cargo.lock]: And we now depend on `zcash_proofs 0.12.1` which removes these duplicates.

**Review by nuttycom — APPROVED** [review]
utACK. I have not individually verified package hashes.

**Pull request merged**
